boards/shared/Inc/Io/Io_SharedAdc.h

```diff
@@ -1,67 +1,55 @@
-// during the manufacturing process
-#define VREFINT_CALIBRATION_ADDRESS		(uint16_t *)(0x1FFFF7BA)
-
-// Number of microcontroller pins that are configured to be ADC inputs
-#ifndef NUM_ADC_CHANNELS
-    #error "Please define the number pins configured as ADC inputs using NUM_ADC_CHANNELS!"
+#ifndef MAX_NUM_OF_ADC
+#define MAX_NUM_OF_ADC 1
+#endif
+#ifndef MAX_NUM_OF_ADC_CHANNELS
+#define MAX_NUM_OF_ADC_CHANNELS 16
 #endif
 
-/** @brief Nominal voltage for VDDA, or ADC power supply */
-#define VDDA_VOLTAGE (float32_t)(3.3f)
-
-/** @brief Number of VREFINT channel. VREFINT is the internal voltage reference 
- *         that provides a stable (bandgap) voltage output for the ADC. */
-#define NUM_VREFINT_CHANNEL 1
+struct AdcInput;
 
 /**
- * @brief  Initialize ADC in DMA mode and the correct max ADC value
- * @param  hadc ADC handle
- * @param  vrefint_regular_rank The Regular Rank of VREFINT channel configured
- *         in STM32CubeMX (Starts at 1)
+ * Allocate and initialize an ADC instance with direct memory access
+ *
+ * @note The user must ensure that only one of the two ADCs is converting
+ * VREFINT as it is forbidden to have several ADCs converting
+ * VREFINT at the same time. Create the ADC instance containing VREFINT first
+ * while initializing the ADCs
+ *
+ * @note To set up the given ADC with a DMA controller, it must be initialized
+ * with (1) Continuous Conversion Mode enabled,(2) DMA Continuous Requests
+ * enabled, (3) DMA buffer with Circular Mode selected
+ *
+ * @param adc_handle The handle for the given ADC
+ * @param vrefint_rank The rank which is used to indicate the index of VREFINT
+ * @return A pointer to the allocated and initialized ADC instance
  */
-void Io_SharedAdc_StartAdcInDmaMode(
-    ADC_HandleTypeDef *hadc,
-    uint32_t           vrefint_regular_rank);
+struct AdcInput *
+    Io_SharedAdc_Init(ADC_HandleTypeDef *adc_handle, size_t vrefint_rank);
 
 /**
- * @brief  Get the appropriate maximum ADC value based on the ADC resolution
- * @return Maximum ADC value in bits
+ * Gets the ADC voltage of the given channel from the given ADC instance
+ * @param adc_input The given ADC instance to get the voltage from
+ * @param adc_rank The rank of the ADC channel to get the voltage from
+ * @return The voltage of the given ADC handle's given channel
  */
-uint32_t Io_SharedAdc_GetAdcMaxValue(void);
+float Io_SharedAdc_GetChannelVoltage(
+    const struct AdcInput *adc_input,
+    size_t                 adc_rank);
 
 /**
- * @brief  Get the array of ADC readings transferred over DMA request
- * @return Pointer to the array of ADC readings
+ * Gets the pointer to the adc handle from the given ADC instance
+ * @param adc_input The given ADC instance to acquire the adc handle from
+ * @return The ADC handle for the specific ADC instance
  */
-const uint32_t *Io_SharedAdc_GetAdcValues(void);
+ADC_HandleTypeDef *Io_SharedAdc_GetAdcHandle(const struct AdcInput *adc_input);
```

Comment:
What's an actual use case for `Io_SharedAdc_GetAdcHandle` and `Io_SharedAdc_GetNumActiveChannel`?